nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/processors/script/ScriptedRouteRecord.java

```diff
@@ -0,0 +1,177 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.nifi.processors.script;
+
+import org.apache.nifi.annotation.behavior.DynamicRelationship;
+import org.apache.nifi.annotation.documentation.CapabilityDescription;
+import org.apache.nifi.annotation.documentation.SeeAlso;
+import org.apache.nifi.annotation.documentation.Tags;
+import org.apache.nifi.components.PropertyDescriptor;
+import org.apache.nifi.components.ValidationContext;
+import org.apache.nifi.components.ValidationResult;
+import org.apache.nifi.processor.ProcessorInitializationContext;
+import org.apache.nifi.processor.Relationship;
+import org.apache.nifi.processor.util.StandardValidators;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+
+@Tags({"record", "routing", "script", "groovy", "jython", "python", "segment", "split", "group", "organize"})
+@CapabilityDescription(
+        "This processor provides the ability to route the records of the incoming FlowFile using an user-provided script. " +
+        "The script is expected to handle a record as argument and return with a string value. " +
+        "The returned value defines a route. All routes are bounded to an outgoing relationship where the record will be transferred to. " +
+        "Relationships are defined as dynamic properties: dynamic property names are serving as the name of the route. " +
+        "The value of a dynamic property defines the relationship the given record will be routed into. Multiple routes might point to the same relationship. " +
+        "Creation of these dynamic relationship is managed by the processor. " +
+        "The records, which for the script returned with an unknown relationship name are routed to the \"unmatched\" relationship. " +
+        "The records are batched: for an incoming FlowFile, all the records routed towards a given relationship are batched into one single FlowFile."
+)
+@SeeAlso(classNames = {
+    "org.apache.nifi.processors.script.ScriptedTransformRecord",
+    "org.apache.nifi.processors.script.ScriptedPartitionRecord",
+    "org.apache.nifi.processors.script.ScriptedValidateRecord",
+    "org.apache.nifi.processors.script.ScriptedFilterRecord"
+})
+@DynamicRelationship(name = "Name from Dynamic Property", description = "FlowFiles that match the Dynamic Property's Attribute Expression Language")
+public class ScriptedRouteRecord extends ScriptedRouterProcessor<String> {
```

Comment:
I think the other processors are very valuable. But I question the necessity of this Processor. I find the configuration very confusing and not user friendly. The user has to understand the possible values that the script could return, and then the user has to go in and redefine them as property keys and map each of those possible values to a relationship. It feels awkward and creates a tight coupling between a 'partition' and a Relationship.

The more conventional approach would be to just use the partition processor. Then, if they want to route based on that, use RouteOnAttribute. The benefits there are that RouteOnAttribute is far more powerful as it enables the full power of Expression Language, it's already well understood, it provides the ability to inspect the attribute value that will be used for routing, which makes it simpler to use in a dataflow, and it decouples the concepts of routes and partitions.

Reply:
I see the value in doing this splitting and routing in one step. Both it feels "natural" to send records in different "buckets" into different paths and performance wise this might come handy with bigger amount of records. If you are not against it, I would rather keep the processor here. But I also understand your concerns and think them valid, so I see it as a possible solution to provide a much more exhaustive documentation of the processor's behaviour, with examples (in the additionalDetails). How do you feel about this approach?